### PR TITLE
[Clang] CGCoroutines skip emitting try block for value returning `noexcept` init `await_resume` calls

### PR DESCRIPTION
**Pull request by yuxuanchen1997**
Previously we were not properly skipping the generation of the `try { }` block around the `init_suspend.await_resume()` if the `await_resume` is not returning void. The reason being that the resume expression was wrapped in a `CXXBindTemporaryExpr` and the first dyn_cast failed, silently ignoring the noexcept. This only mattered for `init_suspend` because it had its own try block.

This patch changes to first extract the sub expression when we see a `CXXBindTemporaryExpr`. Then perform the same logic to check for `noexcept`.

Another version of this patch also wanted to assert the second step by `cast<CXXMemberCallExpr>` and as far as I understand it should be a valid assumption. I can change to that if upstream prefers.

### CODE REVIEW

File: clang/lib/CodeGen/CGCoroutine.cpp [hunk 1]
```diff
@@ -12,9 +12,10 @@
 
 #include "CGCleanup.h"
 #include "CodeGenFunction.h"
-#include "llvm/ADT/ScopeExit.h"
+#include "clang/AST/ExprCXX.h"
 #include "clang/AST/StmtCXX.h"
 #include "clang/AST/StmtVisitor.h"
+#include "llvm/ADT/ScopeExit.h"
```
Comment (ChuanqiXu9) [inline clang/lib/CodeGen/CGCoroutine.cpp:18]: Is this change necessary?

File: clang/lib/CodeGen/CGCoroutine.cpp [hunk 2]
```diff
@@ -129,7 +130,14 @@ static SmallString<32> buildSuspendPrefixStr(CGCoroData &Coro, AwaitKind Kind) {
...
 
-static bool memberCallExpressionCanThrow(const Expr *E) {
+static bool ResumeExprCanThrow(const CoroutineSuspendExpr &S) {
+  const Expr *E = S.getResumeExpr();
+
+  // If the return type of await_resume is not void, get the CXXMemberCallExpr
+  // from its subexpr.
+  if (const auto *BindTempExpr = dyn_cast<CXXBindTemporaryExpr>(E)) {
+    E = BindTempExpr->getSubExpr();
+  }
```
Comment (ChuanqiXu9) [inline clang/lib/CodeGen/CGCoroutine.cpp:140]: Such pattern match doesn't smell good. How about looking into its children recursively if we find `E` is not CXXMemberCallExpr?

File: clang/test/CodeGenCoroutines/coro-init-await-nontrivial-return.cpp [hunk 3]
```diff
@@ -38,9 +39,52 @@ Task coro_create() {
...
+// CHECK-NEXT: call void @_ZN9can_throw14NontrivialTypeD1Ev(
 // CHECK-NEXT: store i1 false, ptr {{.*}}
+}
+
+namespace no_throw {
+struct NontrivialType {
+  ~NontrivialType() {}
+};
+
+struct Task {
```
Comment (ChuanqiXu9) [inline clang/test/CodeGenCoroutines/coro-init-await-nontrivial-return.cpp:55]: It looks a little bit confusing. Let's try to rename it to InitNoThrowTask.

File: clang/lib/CodeGen/CGCoroutine.cpp (hunk 2 above)
Comment (yuxuanchen1997) [reply clang/lib/CodeGen/CGCoroutine.cpp:140]: I find this a little overkill for what we need. The old code here didn't intend to handle more cases either because it just expected the AST to be in a specific shape (`CXXMemberCallExpr`). This fix is merely just to amend its functionality to work on a temporary expression as well should there be a return value to be discarded. It should be clear intent here that we don't care about anything else. (Ideally, I want an assertion here that it's either of the two `Expr` types)

File: clang/test/CodeGenCoroutines/coro-init-await-nontrivial-return.cpp (hunk 3 above)
Comment (yuxuanchen1997) [reply clang/test/CodeGenCoroutines/coro-init-await-nontrivial-return.cpp]: I applied your suggested change. I think this name is still bikesheddable though.

File: clang/lib/CodeGen/CGCoroutine.cpp (hunk 2 above)
Comment (apolloww) [reply clang/lib/CodeGen/CGCoroutine.cpp:139]: I think we are only targeting the resume part of `init_suspend` here, so the cases are limited. Can we change the function name to something more specific, like `InitResumeExprCanThrow`?

Comment (apolloww) [reply clang/lib/CodeGen/CGCoroutine.cpp:139]: I think we are only targeting the resume part of init_suspend here, so the cases are limited. Can we change the function name to something more specific, like `InitResumeExprCanThrow`?

Comment (ChuanqiXu9) [reply clang/lib/CodeGen/CGCoroutine.cpp:139]: Maybe it is not a such big task as you imaged, you can take a look at `Sema::checkFinalSuspendNoThrow` and we should be able to make a simpler change than that.

The reason why I don't like the current pattern is that, every time I wrote:

```
if (auto *WantedExpressionType = isa<AExpreType>(E))
     E = WantedExpressionType->subExpr();
...
```

I found I always missed some situations.

Comment (yuxuanchen1997) [reply clang/lib/CodeGen/CGCoroutine.cpp]: I guess you are correct that this is easy to miss some cases. In my case I missed the fact that the destructor of the `BindTempExpr->getTemporary()` can throw if marked `noexcept(false)`. I updated the PR to cover this case and including corresponding tests. 

Next up I will try to see how many cases I'll need to cover in a recursive check on `children()` like the one in `Sema::checkFinalSuspendNoThrow`.

**Comment by yuxuanchen1997** [comment]
@ChuanqiXu9, I updated the check like you suggested, and it should look much nicer now. The only caveat is that the `children()` call isn't going to cover the implicit destructor call in a `CXXBindTemporaryExpr`. That pattern matching case is here to stay and likely there are many other types implicit calls that make this analysis unsound (See https://eel.is/c++draft/except.spec#6). In reality none of them can reach here though.

File: clang/lib/CodeGen/CGCoroutine.cpp [hunk 4]
```diff
@@ -129,14 +129,48 @@ static SmallString<32> buildSuspendPrefixStr(CGCoroData &Coro, AwaitKind Kind) {
...
+
+static bool ResumeStmtCanThrow(const Stmt *S) {
+  if (const auto *CE = dyn_cast<CallExpr>(S)) {
+    const auto *Callee = CE->getDirectCallee();
+    if (!Callee)
+      // We don't have direct callee. Conservatively assume throwing.
+      return true;
+
+    if (FunctionCanThrow(Callee))
+      return true;
```
Comment (ChuanqiXu9) [inline clang/lib/CodeGen/CGCoroutine.cpp:152]:
```suggestion
    if (!FunctionCanThrow(Callee))
      return false;
```

nit: This reads better. In case we know the called function is nothrow, we don't need to check it further recursively especially await_resume doesn't have arguments except `this`.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM with nit.

File: clang/lib/CodeGen/CGCoroutine.cpp (hunk 4 above)
Comment (yuxuanchen1997) [reply clang/lib/CodeGen/CGCoroutine.cpp:152]: The logic doesn't go this way. 

If a function can throw, the task of this (conservative) analysis is to return true and nothing else needs to be done. Otherwise, fall through and recursively visit all children, which may include Stmts that throw.

Comment (ChuanqiXu9) [reply clang/lib/CodeGen/CGCoroutine.cpp:152]: Please don't land it when you have different opinions.

It should be safe to return false directly if we know the function is nothrow. Then it may be slightly more efficient.
